1.8 documentation updates: Concepts & menu structure

tgraf opened the pull request.

joestringer commented on Documentation/concepts/ipam/cluster-pool.rst:
```diff
@@ -6,13 +6,24 @@
-see :ref:`k8s_hostscope`. The functionality is the same but the ``PodCIDRs`` are
-allocated and managed entirely by Cilium Operator.
+The Cluster scope IPAM mode is almost identical to the :ref:`k8s_hostscope`
+mode. It also works based on a PodCIDR assigned to each node which is then used
```
Should we start by a single-sentence description of what this is and only mention that it's similar to hostscope later on in the paragraph? Starting with a comparison to hostscope requires the reader to already understand hostscope.

joestringer commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,309 @@
+=============
+
+When no configuration is provided, Cilium automatically runs in this mode as it
+is the mode with the least requirements on the underlying networking
```
```suggestion
is the mode with the fewest requirements on the underlying networking
```

joestringer commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,309 @@
+your choice (`VXLAN` by default). It's up to the overlay implementation and
+the kernel's routing table to route traffic to the other nodes' overlays. 
+It's likely that the encapsulation protocol will pick up the default layer
+3 infrastructure for the host. 
```
I'm not quite sure what this last sentence means. Is it trying to describe that it will likely follow the default route?

tgraf replied:
It was definitely written in a way that is not clear. I've rewritten the entire section.

joestringer commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,309 @@
+If you would like all traffic in cilium tunnels to go through a specific 
+network interface (e.g.: a private network for security purposes), 
+you should configure the host's default endpoint accordingly 
+(e.g.: the IP of the node in a Kubernetes setup). You may check 
```
Can we link an upstream k8s doc for this or does it differ based on the k8s installer being used?

tgraf replied:
I've removed this section entirely.

joestringer commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,309 @@
+
+.. code-block: yaml
+
+        ipam: eni
+        blacklist-conflicting-routes: "false"
+        enable-endpoint-routes: "true"
+        auto-create-cilium-node-resource: "true"
+        egress-masquerade-interfaces: eth+
```
Two thoughts here:

(1) Why can't we provide a global "ENI mode" setting that selects these micro settings automatically?

(2) As far as I know, `egress-masquerade-interfaces` is currently targeted to only work with the iptables-based masquerade and for kube-proxy-free will be superseded by `--devices: ...`. Not sure how we want to handle this potential mismatch. /cc @brb

tgraf replied:
> (1) Why can't we provide a global "ENI mode" setting that selects these micro settings automatically?

This is what we had. The helm chart became very complex and we introduced a severe bug that broke ENI and Azure mode. I'm open to re-add it but not using the complexity we had in the end.

We should discuss this when we decide how to work with Helm in general. We need some consistency.

joestringer reviewed (changes requested):
Some minor comments as a first pass.

christarazi commented on Documentation/concepts/ipam/deep_dive.rst:
```diff
@@ -0,0 +1,19 @@
+
+The control flow picture below gives an overview about how the containers
+obtain its IP Address from the IPAM for different modes of Address Management
+that Cilium Supports.
```
```suggestion
The control flow diagram below gives an overview on how containers
obtain their IP address from the IPAM for each different mode of Address Management
that Cilium Supports.
```

christarazi commented on Documentation/concepts/ipam/index.rst:
```diff
@@ -6,35 +6,24 @@
+IP Address Management (IPAM) is responsible for the allocation and management
+of IP addresses used by network endpoints (container and others) managed by
+Cilium. Various address management modes are supported to the meet the needs of
+different users:
```
Nit: 
```suggestion
Cilium. Various IPAM modes are supported to the meet the needs of
different users:
```
I find that referring to the topic at hand by its explicit name and proper noun ("IPAM" here) is clearer, than the improper noun, "address management"

christarazi commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -0,0 +1,55 @@
+============
+
+IPv4 addresses used for pods are typically allocated from RFC1918 private
+address blocks and are thus not publicly routable. Cilium will automatically
```
```suggestion
address blocks and thus, not publicly routable. Cilium will automatically
```

christarazi commented on Documentation/concepts/networking/masquerading.rst:
```diff
@@ -0,0 +1,55 @@
+  10.0.0.0/8`` in which case all destinations within that CIDR will **not** be
+  masqueraded.
+
+Setting the masquerading interface (only available in iptables mode)
```
Would it help to expand slightly on what "iptables mode" means here in this context?

christarazi replied:
Ah, I see that it is explained a bit later below. I think it would be helpful to add a reference here. Something like:

```suggestion
Setting the masquerading interface (only available in iptables mode [see implementation modes below])
```

christarazi commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,309 @@
+It's likely that the encapsulation protocol will pick up the default layer
+3 infrastructure for the host. 
+
+If you would like all traffic in cilium tunnels to go through a specific 
```
Nit: 
```suggestion
If you would like all traffic in Cilium tunnels to go through a specific 
```

tgraf replied:
I've removed this section

christarazi commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,309 @@
+network interface (e.g.: a private network for security purposes), 
+you should configure the host's default endpoint accordingly 
+(e.g.: the IP of the node in a Kubernetes setup). You may check 
+the endpoints used by cilium to set up the tunneling with
```
```suggestion
the endpoints used by Cilium to set up the tunneling with
```

tgraf replied:
This section was removed

christarazi commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,309 @@
+
+In native routing mode, Cilium will delegate all packets forwarding which are
+not addressed to another local endpoint to the routing subsystem of the Linux
+kernel. This means that the packet will be routed as if a local process would
```
Would it be accurate to say this instead?

```suggestion
In native routing mode, Cilium will delegate all packets which are
not addressed to another local endpoint to the routing subsystem of the Linux
kernel. This means that the packet will be routed as if a local process would
```

christarazi commented on Documentation/concepts/overview.rst:
```diff
@@ -11,97 +11,81 @@ Component Overview
+  The graphical user interface (``hubble-ui``) utilizes relay-based visibility
+  to provide a graphical service dependency and connectivity map.
+
+eBPF
```
:bee: :tada:

christarazi commented on Documentation/concepts/overview.rst:
```diff
@@ -11,97 +11,81 @@ Component Overview
+  visibility & monitoring requirements.
+
+  The Cilium agent listens for events from orchestration systems such as
+  Kubernetes to learn when containers or workloads are started or stopped. It
```
```suggestion
  Kubernetes to learn when containers or workloads are started and stopped. It
```

christarazi commented on Documentation/concepts/overview.rst:
```diff
@@ -11,97 +11,81 @@ Component Overview
+
+Client (CLI)
+  The Cilium CLI client (``cilium``) is a command-line tool that is installed
+  along with the Cilium Agent. It interacts with the REST API of the Cilium
```
The agent has been referred to by its lower-case counterpart several times in this paragraph and the paragraph before it. I'd suggest we stick to one capitalization or the other.

christarazi reviewed (commented):
First-pass comments, looks good overall

tgraf committed:
docs: Restructure IPAM section

* Move the detailed flow deep dive into a sub section
* Mark Cluster Scope / Pool as the default IPAM
* Mark old host scope as legacy

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Add Hubble to introduction chapter

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Rework component overview

* Mention all involved components
* Use the compoonent overview for a high-level description. More
  detailed descriptions will be added in dedicated concepts chapters.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Remove Mesos/Marathon guide

This guide has been unused for a while and is outdated.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Improve cluster pool IPAM documentation

* Improve structure
* Add overview diagram

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Move envoy section under proxy injection in concepts

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Move Kubernetes integration into concepts

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Move remaining architecture topics into concepts chapter

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Split network security chapter into multiple files

Also removes the oudated Kubernetes section.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Remove docker integration

The information is duplicated from the getting started guide.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Remove Istio integration section

It only refers to the getting started guide

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Move IPAM section into networking chapter

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Rename security to network security in concepts

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Consolidate networking section about routing

The different routing modes have been spread across multiple different
chapters. Consolidate it all in a single place: Concepts -> Networking
-> Routing.

Also split out multi-cluster to extend it later.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Consolidate Administration and Configuration into Operations

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Move scalability report into concepts

Signed-off-by: Thomas Graf <thomas@cilium.io>

michi-covalent commented on Documentation/intro.rst:
```diff
@@ -23,8 +23,66 @@ within Linux itself.  Because BPF runs inside the Linux kernel, Cilium
+* Is any network communication failing? Why is communication failing? Is it
+  DNS? Is it an application or network problem? Is the communication broken on
+  layer 4 (TCP) or layer 7 (HTTP)?
+* Which services have experienced a DNS resolution problems in the last 5
```
```suggestion
* Which services have experienced a DNS resolution problem in the last 5
```

michi-covalent commented on Documentation/intro.rst:
```diff
@@ -23,8 +23,66 @@ within Linux itself.  Because BPF runs inside the Linux kernel, Cilium
+eBPF is enabling visibility into and control over systems and applications at a
+granularity and efficiency that was not possible before. It does so in a
+completely transparent way, without requiring the application to change in any
+way. eBPF equally well-equipped to handle modern containerized workloads as
```
```suggestion
way. eBPF is equally well-equipped to handle modern containerized workloads as
```

michi-covalent commented on Documentation/concepts/overview.rst:
```diff
@@ -11,97 +11,81 @@ Component Overview
+Hubble
+======
+
+Agent
```
kind of feel like we should pick one word, either agent or server, but not both. i'm slightly inclined to call it Hubble Server, but i can go either way.

michi-covalent replied:
kind of feel like we should pick one word, either agent or server, but not both. i'm slightly inclined to call it Hubble Server, but i can go either way.

michi-covalent reviewed (commented):
reviewed:

- Documentation/intro.rst
- Documentation/index.rst
- Documentation/concepts/overview.rst

qmonnet commented on Documentation/concepts/ipam/index.rst:
```diff
@@ -6,35 +6,24 @@
-look at overview of the cilium container networking control flow.
+IP Address Management (IPAM) is responsible for the allocation and management
+of IP addresses used by network endpoints (container and others) managed by
+Cilium. Various IPAM modes are supported to the meet the needs of different
```
```suggestion
Cilium. Various IPAM modes are supported to meet the needs of different
```

qmonnet commented on Documentation/gettingstarted/mesos.rst:
```diff
@@ -1,359 +0,0 @@
-
-**L3/L4 Policy with Cilium and Mesos**
-
-.. image:: images/cilium_mesos_demo_l3-l4-policy-170817.png
```
This picture is no longer used and can probably be removed.

qmonnet commented on Documentation/gettingstarted/mesos.rst:
```diff
@@ -1,359 +0,0 @@
-
-**L7 Policy with Cilium and Mesos**
-
-.. image:: images/cilium_mesos_demo_l7-policy-230817.png
```
This one could also be removed.

qmonnet commented on Documentation/concepts/overview.rst:
```diff
@@ -11,97 +11,81 @@ Component Overview
+compiler converts the bytecode to CPU architecture specific instructions for
+native execution efficiency. BPF programs can be run at various hooking points
+in the kernel such as for incoming packets, outgoing packets, system calls,
+kprobes, etc.
```
Nit: I'd remove `system calls` and `kprobes` from this paragraph (or mention they are used for tracing use cases). Since the rest of it the document is dealing with networking and Cilium concepts, it might be confusing if readers e.g. do not know what kprobes are but feel it's a concept they need to understand.

---

Other nit, not related, at the end of this section (I can't comment directly on it):
```
Linux distros that focus on being a container runtime (e.g., CoreOS, Fedora
Atomic) typically already ship kernels that are newer than 4.8, but even recent
versions of general purpose operating systems such as Ubuntu 16.10 ship fairly
recent kernels. Some Linux distributions still ship older kernels but many of
them allow installing recent kernels from separate kernel package repositories.
```
Does Ubuntu 16.10 still qualify as “recent” or should this need an update?

tgraf replied:
I've removed the entire paragraph as it's becoming an exception that 4.9 can't be met.

qmonnet commented on Documentation/intro.rst:
```diff
@@ -23,8 +23,66 @@ within Linux itself.  Because BPF runs inside the Linux kernel, Cilium
+===============
+
+Hubble is a fully distributed networking and security observability platform.
+It is built on top of Cilium and eBPF to enable deep visibility into the
```
`What is Cilium?` uses `BPF`, all new contents have `eBPF`. Could you update the Cilium intro for consistency, please?

qmonnet commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,319 @@
+  Encapsulation protocols allow for the carrying of metadata along with the
+  network packet. Cilium makes use of this ability to transfer metadata such as
+  the source security identity and load balancing state to perform
+  direct-server-return.
```
```suggestion
  direct-server-return (DSR).
```

qmonnet commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,319 @@
+  Due to adding encapsulation headers, the effective MTU available for payload
+  is lower than with native-routing (50 bytes per network packet for VXLAN).
+  This results in a lower maximum throughput rate for a particular network
+  connection. This can be largely mitigated by enabling jumbo frames (50 bytes
+  of overhead for each 1500 bytes vs 50 bytes of overhead for each 9000 bytes).
```
How does this play with XDP acceleration? As far as I know it does not work with jumbo frames yet?

tgraf replied:
I'm not sure what the question or proposal is. How does XDP play a role with encapsulation?

qmonnet commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,319 @@
+routing mode:
+
+* ``tunnel: disabled``: Enable native routing mode
+* ``enable-endpoint-routes: true``: Enable per-endpoint routing on the node
```
```suggestion
* ``tunnel: disabled``: Enable native routing mode.
* ``enable-endpoint-routes: true``: Enable per-endpoint routing on the node.
```

qmonnet commented on Documentation/concepts/networking/routing.rst:
```diff
@@ -0,0 +1,319 @@
+The following configuration options must be set to run the datapath on GKE:
+
+* ``gke.enabled: true``: Enables the Google Kubernetes Engine (GKE) datapath.
+  Setting this to true will enabled the following options:
```
```suggestion
  Setting this to ``true`` will enable the following options:
```

qmonnet commented on Documentation/concepts/ebpf/iptables.rst:
```diff
@@ -0,0 +1,23 @@
+Iptables Usage
+##############
+
+Depending on the Linux kernel version being used. The eBPF datapath can
```
```suggestion
Depending on the Linux kernel version being used, the eBPF datapath can
```

qmonnet commented on Documentation/concepts/ebpf/lifeofapacket.rst:
```diff
@@ -0,0 +1,100 @@
+
+################
+Life of a Packet
+################
```
Unless I am mistaken, all the contents you add in this file and the other ones created in this commit come as duplicates with the `concepts/ebpf/intro.rst` (formerly `architecture.rst`). I think you intended to move it but forgot to remove the original text?

qmonnet commented on Documentation/concepts/ebpf/lifeofapacket.rst:
```diff
@@ -0,0 +1,100 @@
+needs to be re-traversed when handing the packet from one veth device to its
+peer device in the other network namespace. This egress-to-ingress switch needs
+to be done twice when communicating between local Cilium endpoints, and once
+for packet that are arriving or sent out of the host.
```
```suggestion
for packets that are arriving or sent out of the host.
```

qmonnet commented on Documentation/concepts/security/intro.rst:
```diff
@@ -0,0 +1,37 @@
+
+*****
+Intro
+*****
```
```suggestion
************
Introduction
************
```

qmonnet commented on Documentation/concepts/security/intro.rst:
```diff
@@ -0,0 +1,37 @@
+  endpoint with the label ``role=backend`` must have a particular security
+  certificate to authenticate itself before being able to initiate any
+  connections. See `GH issue 502
+  <https://github.com/cilium/cilium/issues/502>`_ for additional details.
```
#502 is closed and refers to #504 instead.

tgraf committed:
doc: Fix various typos and grammar issues

Also removes an outdated section about future roadmap

Signed-off-by: Thomas Graf <thomas@cilium.io>
